express_relay/programs/express_relay/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+        validate_fee_split(data.split_protocol)?;
+
+        ctx.accounts.protocol_config.split = data.split_protocol;
+
+        Ok(())
+    }
+
+    pub fn permission(ctx: Context<Permission>, data: PermissionArgs) -> Result<()> {
+        if data.deadline < Clock::get()?.unix_timestamp as u64 {
+            return Err(ExpressRelayError::DeadlinePassed.into());
+        }
+
+        // check that not cpi
+        let instruction = get_instruction_relative(0, &ctx.accounts.sysvar_instructions.to_account_info())?;
+        if instruction.program_id != crate::id() {
+            return err!(ExpressRelayError::InvalidCPIPermission);
+        }
+
+        // handle bid payment
+        let express_relay_metadata = &ctx.accounts.express_relay_metadata;
+        let protocol_config = &ctx.accounts.protocol_config;
+        let searcher = &ctx.accounts.searcher;
+        let protocol_fee_receiver = &ctx.accounts.fee_receiver_protocol;
+        let relayer_fee_receiver = &ctx.accounts.fee_receiver_relayer;
+
+        let bid_amount = data.bid_amount;
+        let split_relayer = express_relay_metadata.split_relayer;
+        let split_protocol_default = express_relay_metadata.split_protocol_default;
+        let split_protocol: u64;
+
+        let rent_searcher = Rent::default().minimum_balance(0).max(1);
+        if bid_amount + rent_searcher > searcher.to_account_info().lamports() {
+            return err!(ExpressRelayError::InsufficientSearcherFunds);
+        }
+
+        let protocol_config_account_info = protocol_config.to_account_info();
+        if protocol_config_account_info.data_len() > 0 {
+            let account_data = &mut &**protocol_config_account_info.try_borrow_data()?;
+            let protocol_config_data = ConfigProtocol::try_deserialize(account_data)?;
+            split_protocol = protocol_config_data.split;
+        } else {
+            split_protocol = split_protocol_default;
+        }
+
+        let fee_protocol = bid_amount * split_protocol / FEE_SPLIT_PRECISION;
+        if fee_protocol > bid_amount {
+            return err!(ExpressRelayError::FeesTooHigh);
+        }
+
+        let fee_relayer = bid_amount.saturating_sub(fee_protocol) * split_relayer / FEE_SPLIT_PRECISION;
+        if fee_relayer.checked_add(fee_protocol).unwrap() > bid_amount {
+            return err!(ExpressRelayError::FeesTooHigh);
+        }
+
+        let balance_protocol_fee_receiver = protocol_fee_receiver.to_account_info().lamports();
+        // TODO: use actual protocol pda datalen here?
+        let rent_protocol_fee_receiver = Rent::default().minimum_balance(0).max(1);
+        let amount_protocol: u64;
+        if balance_protocol_fee_receiver >= rent_protocol_fee_receiver {
+            amount_protocol = fee_protocol;
+        } else {
+            amount_protocol = fee_protocol + (rent_protocol_fee_receiver - balance_protocol_fee_receiver);
```

Comment:
this would fail if the bid amount can not cover the rent amounts. I think it's ok to be dumber here and just check if this account is already initialized and owned by the protocol. It also forces the protocol to have some instructions for initializing this account and hopefully, for withdrawing fees too

Reply:
i agree. there is a chance we may want to make it possible for `fee_receiver_protocol` to be an arbitrary address (e.g. for frontends), but not 100% sure yet. will leave it as a specified pda for now and we can change it in the future if needed